internal/primitives/runtime/interfaces.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package runtime
+
+import (
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hashing"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"golang.org/x/exp/constraints"
+)
+
+// Number is the header number type
+type Number interface {
+	~uint | ~uint32 | ~uint64
+}
+
+// Hash type
+type Hash interface {
+	constraints.Ordered
+	// Bytes returns a byte slice representation of Hash
+	Bytes() []byte
+	// String returns a unique string representation of the hash
+	String() string
+}
+
+// Hasher is an interface around hashing
+type Hasher[H Hash] interface {
+	// Produce the hash of some byte-slice.
+	Hash(s []byte) H
+
+	// Produce the hash of some codec-encodable value.
+	HashOf(s any) H
+}
+
+// Blake2-256 Hash implementation.
+type BlakeTwo256 struct{}
+
+// Produce the hash of some byte-slice.
+func (bt256 BlakeTwo256) Hash(s []byte) hash.H256 {
+	h := hashing.BlakeTwo256(s)
+	return hash.H256(h[:])
+}
+
+// Produce the hash of some codec-encodable value.
+func (bt256 BlakeTwo256) HashOf(s any) hash.H256 {
+	bytes := scale.MustMarshal(s)
+	return bt256.Hash(bytes)
+}
+
+var _ Hasher[hash.H256] = BlakeTwo256{}
+
+// Header is the interface for a header. It has types for a `Number`,
+// and `Hash`. It provides access to an `ExtrinsicsRoot`, `StateRoot` and
+// `ParentHash`, as well as a `Digest` and a block `Number`.
+type Header[N Number, H Hash] interface {
+	// Returns a reference to the header number.
+	Number() N
```

Comment:
Why this can't be just uint? 
I can understand that theoretically using uint32 for smaller numbers over uint64 saves us some memory, but this is a very vague memory efficiency tunning in trade of code readability. 
But the efficiency costs of having multiple types utterly dwarfs the efficiency gains.

Reply:
In our SCALE package `uint`, `uint32`, and `uint64` are all encoded differently.  We denote `int` and `uint` for the analagous `Compact<u32>` in the `parity-scale-codec`.  The idea of the interface constraint is to support all supported number types, but in practice the relay chains use `uint32/u32`.  Which would mean they would have update Kusama to use `uint64` in ~700 years 😝

Reply:
see https://github.com/ChainSafe/gossamer/pull/4015#discussion_r1648183205

I'd also like to note that introduction of generics for this feature as well as the GRANDPA work is an explicit choice.  I'd like to be able support the same types that substrate supports, and not just what they run on the relay chains.  We already have relay chain specific implementations of these types (Header, Block, etc) in the Gossamer codebase, but this will replace that.

Reply:
But do we need to bring all the Substrate complexity to gossamer assuming that we are not building Framework?

Reply:
Ideally for the block and state persistence layer we support the interfaces/traits for block types and hashing. I don't agree with the notion that introducing interfaces introduces added complexity. These interfaces and constraints are clearly defined.  This wasn't possible before go generics.